### PR TITLE
Adding Synthetics Capability to Application Signals MCP server

### PR DESCRIPTION
**Pull request by rgupta2205**
Add CloudWatch Synthetics Canary Analysis to Application Signals MCP Server

### Changes

Extended the CloudWatch Application Signals MCP Server with comprehensive CloudWatch Synthetics canary analysis capabilities:

- **New `canary_utils.py`**: Utility functions for canary failure analysis, IAM role validation, artifact retrieval, and debugging
- **Enhanced `server.py`**: Added `analyze_canary_failures` tool for comprehensive canary failure investigation
- **Updated `aws_clients.py`**: Added Synthetics, S3, IAM, Lambda, and STS clients; cleaned up debug logging
- **Updated `README.md`**: Added documentation for canary analysis capabilities
- **Added comprehensive tests**: 21 test scenarios covering real-world canary failure patterns in `test_canary_utils.py`

### User experience

**Before**: Users could only analyze Application Signals metrics and traces, with no visibility into CloudWatch Synthetics canary failures

**After**: Users can now:
- Analyze canary failures with comprehensive root cause identification
- Check IAM role configurations and permissions for canaries
- Retrieve and analyze canary artifacts (logs, screenshots, HAR files)
- Get actionable recommendations for resolving canary issues
- Correlate canary failures with Application Signals telemetry data
- Handle edge cases like S3 access errors, connection timeouts, browser target issues, and VPC configuration problems

### Key Features Added

- **Comprehensive failure analysis** with pattern recognition for common canary issues
- **IAM role and policy validation** with specific permission checks
- **Artifact analysis** including HAR files, screenshots, and logs
- **APM Pulse integration** for enhanced diagnostics
- **21 comprehensive test scenarios** covering real-world failure patterns
- **Robust error handling** for AWS API failures and edge cases

## Checklist

* [x] I have reviewed the [contributing guidelines](https://github.com/awslabs/mcp/blob/main/CONTRIBUTING.md)
* [x] I have performed a self-review of this change
* [x] Changes have been tested (21/21 tests passing)
* [x] Changes are documented

Is this a breaking change? (Y/N) **N**

**RFC issue number**: N/A

## Acknowledgment

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of the [project license](https://github.com/awslabs/mcp/blob/main/LICENSE). 

**Reference**: https://quip-amazon.com/J1eLAngUZYMB/Cloudwatch-Synthetics-MCP-Server-using-API

### CODE REVIEW

File: src/cloudwatch-appsignals-mcp-server/awslabs/cloudwatch_appsignals_mcp_server/canary_utils.py [hunk 1]
```diff
@@ -0,0 +1,827 @@
...
+            import os
+            import ssl
+            
+            layer_arn = custom_layers[0]['Arn']
+            layer_response = lambda_client.get_layer_version_by_arn(Arn=layer_arn)
+            
+            if 'Location' in layer_response['Content']:
+                layer_url = layer_response['Content']['Location']
+                
+                ssl_context = ssl._create_unverified_context()
```
Comment (yiyuan-he) [inline src/cloudwatch-appsignals-mcp-server/awslabs/cloudwatch_appsignals_mcp_server/canary_utils.py]: Is there a specific reason we're doing this? I believe this is disabling SSL certificate verification which should be a security vulnerability.

File: src/cloudwatch-appsignals-mcp-server/awslabs/cloudwatch_appsignals_mcp_server/canary_utils.py [hunk 2]
```diff
@@ -0,0 +1,827 @@
...
+        #         }
+        #     }],
+        #     'Auditors': ['canary', 'operation_metric', 'trace']
+        # }
+        # return await execute_audit_api(audit_input, region, f"Canary Analysis for {canary_name}\n")
+
+        cmd = [
+            'aws', 'application-signals-demo', 'list-audit-findings',
+            '--region', region,
+            '--endpoint-url', 'https://application-signals-gamma.us-west-2.api.aws',
```
Comment (yiyuan-he) [inline src/cloudwatch-appsignals-mcp-server/awslabs/cloudwatch_appsignals_mcp_server/canary_utils.py]: Why are we hardcoding to gamma endpoint? What is the long-term strategy?

File: src/cloudwatch-appsignals-mcp-server/awslabs/cloudwatch_appsignals_mcp_server/canary_utils.py [hunk 3]
```diff
@@ -0,0 +1,827 @@
...
+    
+    role_name = execution_role_arn.split('/')[-1]
+    
+    try:
+        policies_response = iam_client.list_attached_role_policies(RoleName=role_name)
+        attached_policies = policies_response['AttachedPolicies']
+        
+        canary_bucket = canary.get('ArtifactS3Location', '')
+        
+        if not canary_bucket.startswith('s3://'):
```
Comment (yiyuan-he) [inline src/cloudwatch-appsignals-mcp-server/awslabs/cloudwatch_appsignals_mcp_server/canary_utils.py:506]: Do we need to validate resources other than S3 buckets in the future?

File: src/cloudwatch-appsignals-mcp-server/awslabs/cloudwatch_appsignals_mcp_server/canary_utils.py [hunk 4]
```diff
@@ -0,0 +1,827 @@
...
+            endTime=int(end_time.timestamp()),
+            queryString=query
+        )
+        
+        query_id = response['queryId']
+        
+        # Wait for completion
+        max_wait = 30
+        wait_time = 0
+        while wait_time < max_wait:
```
Comment (yiyuan-he) [inline src/cloudwatch-appsignals-mcp-server/awslabs/cloudwatch_appsignals_mcp_server/canary_utils.py:715]: Is this polling strategy sufficient? Did we consider exponential backoff?

File: src/cloudwatch-appsignals-mcp-server/tests/test_canary_utils.py [hunk 5]
```diff
@@ -0,0 +1,1314 @@
...
+        }
+        
+        # Mock subprocess for APM Pulse
+        mock_subprocess.return_value = MagicMock(
+            returncode=0,
+            stdout='APM Pulse analysis: No critical issues detected'
+        )
+
+        result = await analyze_canary_failures(canary_name='test-canary', region='us-east-1')
+
```
Comment (yiyuan-he) [inline src/cloudwatch-appsignals-mcp-server/tests/test_canary_utils.py:164]: Consider adding some assertions here to validate the mock functions were called correctly

Example:
```
mock_har.assert_called()
mock_screenshots.assert_called()
mock_logs.assert_called()
```

File: src/cloudwatch-appsignals-mcp-server/tests/test_canary_utils.py [hunk 6]
```diff
@@ -0,0 +1,1314 @@
...
+        }
+
+        # Mock subprocess for APM Pulse
+        mock_subprocess.return_value = MagicMock(
+            returncode=0,
+            stdout='APM Pulse analysis: Disk space issues detected'
+        )
+
+        result = await analyze_canary_failures(canary_name='enospc-canary', region='us-east-1')
+
```
Comment (yiyuan-he) [inline src/cloudwatch-appsignals-mcp-server/tests/test_canary_utils.py:457]: Verify disk usage metrics were extracted
```
mock_disk_usage.assert_called_once_with('enospc-canary', 'us-east-1')
```

File: src/cloudwatch-appsignals-mcp-server/tests/test_canary_utils.py [hunk 7]
```diff
@@ -0,0 +1,1314 @@
...
+        }
+
+        # Mock subprocess for APM Pulse
+        mock_subprocess.return_value = MagicMock(
+            returncode=0,
+            stdout='APM Pulse analysis: S3 access issues detected'
+        )
+
+        result = await analyze_canary_failures(canary_name='s3-error-canary', region='us-east-1')
+
```
Comment (yiyuan-he) [inline src/cloudwatch-appsignals-mcp-server/tests/test_canary_utils.py:604]: Verify invoke
```
mock_logs.assert_called_once()
```

File: src/cloudwatch-appsignals-mcp-server/tests/test_canary_utils.py [hunk 8]
```diff
@@ -0,0 +1,1314 @@
...
+            'iam_client': mock_iam_client,
+            'synthetics_client': mock_synthetics_client,
+            's3_client': mock_s3_client,
+            'sts_client': mock_sts_client,
+        }
+    finally:
+        for p in patches:
+            p.stop()
+
+
```
Comment (yiyuan-he) [inline src/cloudwatch-appsignals-mcp-server/tests/test_canary_utils.py:36]: We should also add a test for malformed HAR files

File: src/cloudwatch-appsignals-mcp-server/awslabs/cloudwatch_appsignals_mcp_server/canary_utils.py (hunk 2 above)
Comment (rgupta2205) [reply src/cloudwatch-appsignals-mcp-server/awslabs/cloudwatch_appsignals_mcp_server/canary_utils.py]: No, It's temporary. Will be replacing with ListAuditFindings API call. It's releasing on Monday.

File: src/cloudwatch-appsignals-mcp-server/awslabs/cloudwatch_appsignals_mcp_server/canary_utils.py (hunk 3 above)
Comment (rgupta2205) [reply src/cloudwatch-appsignals-mcp-server/awslabs/cloudwatch_appsignals_mcp_server/canary_utils.py:506]: Canary currently requires access to s3 bucket only. Rest we are handling from ListAuditFindings API.

File: src/cloudwatch-appsignals-mcp-server/tests/test_canary_utils.py (hunk 8 above)
Comment (rgupta2205) [reply src/cloudwatch-appsignals-mcp-server/tests/test_canary_utils.py:36]: Done

File: src/cloudwatch-appsignals-mcp-server/awslabs/cloudwatch_appsignals_mcp_server/canary_utils.py (hunk 4 above)
Comment (rgupta2205) [reply src/cloudwatch-appsignals-mcp-server/awslabs/cloudwatch_appsignals_mcp_server/canary_utils.py:715]: Added exponential backoff.

File: src/cloudwatch-appsignals-mcp-server/awslabs/cloudwatch_appsignals_mcp_server/canary_utils.py (hunk 1 above)
Comment (rgupta2205) [reply src/cloudwatch-appsignals-mcp-server/awslabs/cloudwatch_appsignals_mcp_server/canary_utils.py]: Shift to another approach

File: src/cloudwatch-appsignals-mcp-server/tests/test_canary_utils.py (hunk 6 above)
Comment (rgupta2205) [reply src/cloudwatch-appsignals-mcp-server/tests/test_canary_utils.py:457]: Added

File: src/cloudwatch-appsignals-mcp-server/tests/test_canary_utils.py (hunk 7 above)
Comment (rgupta2205) [reply src/cloudwatch-appsignals-mcp-server/tests/test_canary_utils.py:604]: Added

File: src/cloudwatch-appsignals-mcp-server/tests/test_canary_utils.py (hunk 5 above)
Comment (rgupta2205) [reply src/cloudwatch-appsignals-mcp-server/tests/test_canary_utils.py:164]: Added

File: src/cloudwatch-appsignals-mcp-server/awslabs/cloudwatch_appsignals_mcp_server/aws_clients.py [hunk 9]
```diff
@@ -35,6 +35,11 @@ def _initialize_aws_clients():
     logs_endpoint = os.environ.get('MCP_LOGS_ENDPOINT')
     cloudwatch_endpoint = os.environ.get('MCP_CLOUDWATCH_ENDPOINT')
     xray_endpoint = os.environ.get('MCP_XRAY_ENDPOINT')
+    synthetics_endpoint = os.environ.get('MCP_SYNTHETICS_ENDPOINT')
+    s3_endpoint = os.environ.get('MCP_S3_ENDPOINT')
+    iam_endpoint = os.environ.get('MCP_IAM_ENDPOINT')
+    lambda_endpoint = os.environ.get('MCP_LAMBDA_ENDPOINT')
+    sts_endpoint = os.environ.get('MCP_STS_ENDPOINT')
```
Comment (mxiamxia) [inline src/cloudwatch-appsignals-mcp-server/awslabs/cloudwatch_appsignals_mcp_server/aws_clients.py]: The names of these endpoints are very generic. Using something like MCP_S3_ENDPOINT might make customers think they need to configure a general S3 service endpoint. I assume these are intended only for the Synthetics use case? Also, is specifying them optional, and under what scenarios would customers need to set them?

Comment (mxiamxia) [reply src/cloudwatch-appsignals-mcp-server/awslabs/cloudwatch_appsignals_mcp_server/aws_clients.py]: We don’t want customers to have to specify any of these endpoints. The few we have for CloudWatch are only for internal testing. I suggest keeping only MCP_SYNTHETICS_ENDPOINT if need and removing the others; otherwise, we’d have to iterate all service endpoints as we expand.

File: src/cloudwatch-appsignals-mcp-server/awslabs/cloudwatch_appsignals_mcp_server/canary_utils.py [hunk 10]
```diff
@@ -0,0 +1,822 @@
...
+# limitations under the License.
+
+"""Utility functions for CloudWatch Synthetics canary analysis and debugging."""
+
+import asyncio
+import json
+import gzip
+import io
+import zipfile
+import subprocess
```
[the remark this answers is not in the document]
Comment (mxiamxia) [reply src/cloudwatch-appsignals-mcp-server/awslabs/cloudwatch_appsignals_mcp_server/canary_utils.py]: is this used?

File: src/cloudwatch-appsignals-mcp-server/awslabs/cloudwatch_appsignals_mcp_server/canary_utils.py [hunk 11]
```diff
@@ -0,0 +1,822 @@
...
+import os
+import re
+import urllib.request
+from datetime import datetime, timedelta
+from typing import Optional, Dict, List, Any
+from urllib.parse import urlparse
+from botocore.exceptions import ClientError
+from loguru import logger
+from .aws_clients import logs_client, synthetics_client, s3_client, iam_client, lambda_client, sts_client
+from .utils import parse_timestamp
```
Comment (mxiamxia) [inline src/cloudwatch-appsignals-mcp-server/awslabs/cloudwatch_appsignals_mcp_server/canary_utils.py]: can you cleanup all unused imports

File: src/cloudwatch-appsignals-mcp-server/awslabs/cloudwatch_appsignals_mcp_server/canary_utils.py [hunk 12]
```diff
@@ -0,0 +1,822 @@
...
+        
+        if error_patterns:
+            insights.append(f"🚨 Found {len(error_patterns)} error patterns in logs:")
+            for i, error in enumerate(error_patterns[:5], 1):
+                insights.append(f"  {i}. {error}")
+        elif is_failed_run:
+            insights.append("📋 No obvious error patterns found in log files")
+            insights.append("💡 Check CloudWatch Logs for more detailed error information")
+        
+        insights.append(f"📊 Analyzed {min(len(logs), 3)} log files")
```
Comment (mxiamxia) [inline src/cloudwatch-appsignals-mcp-server/awslabs/cloudwatch_appsignals_mcp_server/canary_utils.py]: why only 3 files? is it magic number? BTW, what could be size of insights appended together? will it blow out context window?

File: src/cloudwatch-appsignals-mcp-server/awslabs/cloudwatch_appsignals_mcp_server/canary_utils.py [hunk 13]
```diff
@@ -0,0 +1,822 @@
...
+        }
+        
+        source_location_arn = canary.get('Code', {}).get('SourceLocationArn', '')
+        if source_location_arn and ':layer:' in source_location_arn:
+            try:
+                layer_response = lambda_client.get_layer_version_by_arn(Arn=source_location_arn)
+                if 'Location' in layer_response['Content']:
+                    with tempfile.NamedTemporaryFile(suffix='.zip', delete=False) as tmp_file:
+                        import requests
+                        response = requests.get(layer_response['Content']['Location'])
```
[the remark this answers is not in the document]
Comment (mxiamxia) [reply src/cloudwatch-appsignals-mcp-server/awslabs/cloudwatch_appsignals_mcp_server/canary_utils.py]: need to setup timeout param for the http request

File: src/cloudwatch-appsignals-mcp-server/awslabs/cloudwatch_appsignals_mcp_server/canary_utils.py [hunk 14]
```diff
@@ -0,0 +1,822 @@
...
+                            if handler:
+                                handler_path = handler.replace('.handler', '')
+                                canary_file = next((f for f in code_files if handler_path in f), None)
+                                if canary_file:
+                                    with zip_ref.open(canary_file) as f:
+                                        code_content = f.read().decode('utf-8')
+                                        lines = code_content.split('\n')
+                                        result["code_content"] = '\n'.join(f"{i+1}: {line}" for i, line in enumerate(lines))
+                        os.unlink(tmp_file.name)
+            except Exception:
```
[the remark this answers is not in the document]
Comment (mxiamxia) [reply src/cloudwatch-appsignals-mcp-server/awslabs/cloudwatch_appsignals_mcp_server/canary_utils.py]: do you need to log the error but just pass it?

File: src/cloudwatch-appsignals-mcp-server/awslabs/cloudwatch_appsignals_mcp_server/canary_utils.py [hunk 15]
```diff
@@ -0,0 +1,822 @@
...
+        if not result["code_content"]:
+            custom_layers = [l for l in config.get('Layers', []) if ':layer:Synthetics' not in l['Arn']]
+            
+            for layer in custom_layers:
+                try:
+                    layer_response = lambda_client.get_layer_version_by_arn(Arn=layer['Arn'])
+                    if 'Location' in layer_response['Content']:
+                        with tempfile.NamedTemporaryFile(suffix='.zip', delete=False) as tmp_file:
+                            import requests
+                            response = requests.get(layer_response['Content']['Location'])
```
[the remark this answers is not in the document]
Comment (mxiamxia) [reply src/cloudwatch-appsignals-mcp-server/awslabs/cloudwatch_appsignals_mcp_server/canary_utils.py]: timeout setting

File: src/cloudwatch-appsignals-mcp-server/awslabs/cloudwatch_appsignals_mcp_server/canary_utils.py [hunk 16]
```diff
@@ -0,0 +1,822 @@
...
+                            with zipfile.ZipFile(tmp_file.name, 'r') as zip_ref:
+                                code_files = [f for f in zip_ref.namelist() if f.endswith(('.js', '.py'))]
+                                if code_files:
+                                    with zip_ref.open(code_files[0]) as f:
+                                        code_content = f.read().decode('utf-8')
+                                        lines = code_content.split('\n')
+                                        result["code_content"] = '\n'.join(f"{i+1}: {line}" for i, line in enumerate(lines))
+                                        break
+                            os.unlink(tmp_file.name)
+                except Exception:
```
[the remark this answers is not in the document]
Comment (mxiamxia) [reply src/cloudwatch-appsignals-mcp-server/awslabs/cloudwatch_appsignals_mcp_server/canary_utils.py]: why this time it is a continue, and no logging again?

File: src/cloudwatch-appsignals-mcp-server/awslabs/cloudwatch_appsignals_mcp_server/canary_utils.py [hunk 17]
```diff
@@ -0,0 +1,822 @@
...
+                    continue
+        
+        # If no code found in layers, try function code directly
+        if not result["code_content"]:
+            try:
+                code_location = function_response['Code']['Location']
+                with tempfile.NamedTemporaryFile(suffix='.zip', delete=False) as tmp_file:
+                    import requests
+                    response = requests.get(code_location)
```
[the remark this answers is not in the document]
Comment (mxiamxia) [reply src/cloudwatch-appsignals-mcp-server/awslabs/cloudwatch_appsignals_mcp_server/canary_utils.py]: timeout

File: src/cloudwatch-appsignals-mcp-server/awslabs/cloudwatch_appsignals_mcp_server/server.py [hunk 18]
```diff
@@ -792,6 +811,427 @@ async def audit_service_operations(
...
+                                    
+                                    result += f"• Failed requests: {failure_har.get('failed_requests', 0)} vs {success_har.get('failed_requests', 0)}\n"
+                                    result += f"• Total requests: {failure_har.get('total_requests', 0)} vs {success_har.get('total_requests', 0)}\n\n"
+                                    
+                                    if failure_har.get('request_details'):
+                                        result += f"🚨 FAILED REQUESTS:\n"
+                                        for req in failure_har['request_details'][:3]:
+                                            result += f"• {req.get('url', 'Unknown')}: {req.get('status', 'Unknown')} ({req.get('time', 0):.1f}ms)\n"
+                            except Exception:
+                                pass
```
Comment (mxiamxia) [inline src/cloudwatch-appsignals-mcp-server/awslabs/cloudwatch_appsignals_mcp_server/server.py]: is it fine to ignore the errors without any footprint?

File: src/cloudwatch-appsignals-mcp-server/awslabs/cloudwatch_appsignals_mcp_server/server.py [hunk 19]
```diff
@@ -792,6 +811,427 @@ async def audit_service_operations(
         return f'Error: {str(e)}'
 
 
+@mcp.tool()
+async def analyze_canary_failures(canary_name: str, region: str = AWS_REGION) -> str:
+    """Comprehensive canary failure analysis with deep dive into issues.
+
+    Use this tool to:
+    - Deep dive into canary failures with root cause identification
+    - Analyze historical patterns and specific incident details
```
Comment (mxiamxia) [inline src/cloudwatch-appsignals-mcp-server/awslabs/cloudwatch_appsignals_mcp_server/server.py]: A general q: it looks like this tool does multiple things and fetch the date from multiple sources. Do you know how much time it would cost in average cases? MCP Tools default timeout is 60s.

File: src/cloudwatch-appsignals-mcp-server/awslabs/cloudwatch_appsignals_mcp_server/canary_utils.py (hunk 12 above)
Comment (rgupta2205) [reply src/cloudwatch-appsignals-mcp-server/awslabs/cloudwatch_appsignals_mcp_server/canary_utils.py]: Consider 3 files parameter based on experiment to prevent context window overflow. No it will not blow out as we are just capturing error or exception message and  also tested 80+ unique canary scenarios it worked well.

File: src/cloudwatch-appsignals-mcp-server/awslabs/cloudwatch_appsignals_mcp_server/server.py (hunk 19 above)
Comment (rgupta2205) [reply src/cloudwatch-appsignals-mcp-server/awslabs/cloudwatch_appsignals_mcp_server/server.py]: That's a good question! We have multiple processes, but they're conditional—the tool only runs what's needed based on the situation. It uses a fallback approach, so even in the worst case, it takes about 35 seconds

File: src/cloudwatch-appsignals-mcp-server/awslabs/cloudwatch_appsignals_mcp_server/aws_clients.py (hunk 9 above)
Comment (rgupta2205) [reply src/cloudwatch-appsignals-mcp-server/awslabs/cloudwatch_appsignals_mcp_server/aws_clients.py]: ack

File: src/cloudwatch-appsignals-mcp-server/awslabs/cloudwatch_appsignals_mcp_server/server.py [hunk 20]
```diff
@@ -792,6 +811,427 @@
...
+                                success_har_files = [a for a in success_artifacts if a['Key'].lower().endswith(('.har', '.har.gz', '.har.html'))]
+                                
+                                if har_files and success_har_files:
+                                    failure_har = await analyze_har_file(s3_client, bucket_name, har_files, is_failed_run=True)
+                                    success_har = await analyze_har_file(s3_client, bucket_name, success_har_files, is_failed_run=False)
+                                    
+                                    result += f"• Failed requests: {failure_har.get('failed_requests', 0)} vs {success_har.get('failed_requests', 0)}\n"
+                                    result += f"• Total requests: {failure_har.get('total_requests', 0)} vs {success_har.get('total_requests', 0)}\n\n"
+                                    
+                                    if failure_har.get('request_details'):
```
[the remark this answers is not in the document]
Comment (rgupta2205) [reply src/cloudwatch-appsignals-mcp-server/awslabs/cloudwatch_appsignals_mcp_server/server.py:1065]: fixed

**Comment by rgupta2205** [comment]
> Why are there so many new changes? Previously the PR was ~2K lines of code changes and now it is approaching 7K

Just adding more unit test as it's (88.99% of diff hit (target 89.48%)) which causing check failure

**Comment by yiyuan-he** [comment]
> Just adding more unit test as it's (88.99% of diff hit (target 89.48%)) which causing check failure

~0.5% additional test coverage results in 5K more lines of code?

**Comment by rgupta2205** [comment]
> > Just adding more unit test as it's (88.99% of diff hit (target 89.48%)) which causing check failure
> 
> ~0.5% additional test coverage results in 5K more lines of code?

Originally it was in 35%

File: src/cloudwatch-appsignals-mcp-server/cloudwatch-appsignals-mcp-server-coverage.xml [hunk 21]
```diff
@@ -0,0 +1,2567 @@
+<?xml version="1.0" ?>
```
Comment (yiyuan-he) [inline src/cloudwatch-appsignals-mcp-server/cloudwatch-appsignals-mcp-server-coverage.xml]: Why are we including this file?

Comment (rgupta2205) [reply src/cloudwatch-appsignals-mcp-server/cloudwatch-appsignals-mcp-server-coverage.xml]: Removed

**Review by mxiamxia — APPROVED** [review]
LGTM now. Thx

**Review by yiyuan-he — APPROVED** [review]
LGTM ty
